src/Microsoft.Extensions.DependencyInjection.Abstractions/IDesignTimeServiceProviderFactory.cs

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// A factory for creating the application's <see cref="IServiceProvider"/> at design time. Implement this interface
+    /// to enable design-time services that need access to the application's services. Design-time services will
+    /// automatically discover implementations of this interface.
+    /// </summary>
+    public interface IDesignTimeServiceProviderFactory
```

Comment:
I'm thinking about how calls into Startup gets their arguments. Basically that's DI too. Though last time I checked it was a lot of code to make it look that way and not extensible. Making that extensible would be a good reason to have both design time something that can be resolved by tools and a discoverable common one that is always used. Such a pattern *could* also use discovery at runtime, but it could again probably just be Main that brings it before creating the host builder with an API call to replace the default or just a call on host builder itself. But going after this will be way bigger a change than envisioned I guess.